static/js/unread_ops.js

```diff
@@ -13,36 +13,69 @@ import * as notifications from "./notifications";
 import * as people from "./people";
 import * as recent_topics_ui from "./recent_topics_ui";
 import * as recent_topics_util from "./recent_topics_util";
-import * as reload from "./reload";
 import * as ui_report from "./ui_report";
 import * as unread from "./unread";
 import * as unread_ui from "./unread_ui";
 
 // anchor message is not included in this calculation
 const NUM_OF_MESSAGES_UPDATED_PER_BATCH = 5000;
 let loading_indicator_displayed = false;
-export function mark_all_as_read() {
-    unread.declare_bankruptcy();
-    unread_ui.update_unread_counts();
+let unread_count_before_marking_all_msgs_read = 0;
+let messages_read_till_now = 0;
 
+export function mark_all_as_read() {
+    if (unread_count_before_marking_all_msgs_read === 0) {
+        unread_count_before_marking_all_msgs_read = unread.get_unread_message_count();
+    }
+    const opts = {
+        anchor: "first_unread",
+        include_anchor: true,
+        num_before: 0,
+        num_after: NUM_OF_MESSAGES_UPDATED_PER_BATCH,
+        op: "add",
+        flag: "read",
+    };
     channel.post({
-        url: "/json/mark_all_as_read",
-        success: () => {
-            // After marking all messages as read, we reload the browser.
-            // This is useful to avoid leaving ourselves deep in the past.
-            // This is also the currently intended behavior in case of partial success,
-            // (response code 200 with result "partially_completed")
-            // where the request times out after marking some messages as read,
-            // so we don't need to distinguish that scenario here.
-            // TODO: The frontend handling of partial success can be improved
-            // by re-running the request in a loop, while showing some status indicator
-            // to the user.
-            reload.initiate({
-                immediate: true,
-                save_pointer: false,
-                save_narrow: true,
-                save_compose: true,
-            });
+        url: "/json/messages/flags",
+        data: opts,
+        success(data) {
+            if (!data.found_newest) {
```

Comment:
Maybe add a block comment here explaining that we're only displaying a progress indicator in the case that there's so many messages to modify that we were not able to modify them all in a single request; if we can process them all in a single request, we expect the user to get immediate feedback in the form of the unread count disappearing.

Reply:
Done!